File: examples/03-supervision/typescript/supervision.ts
/**
 * Supervision Pattern - 감독 트리 예제
 *
 * Actor 장애 처리와 복구 전략을 구현합니다.
 */

// ============================================
// 감독 전략 정의
// ============================================

enum Directive {
  Resume = 'RESUME',      // 상태 유지하고 계속
  Restart = 'RESTART',    // 상태 초기화 후 재시작
  Stop = 'STOP',          // 종료
  Escalate = 'ESCALATE',  // 부모에게 전파
}

type Decider = (error: Error) => Directive;

interface SupervisorStrategy {
  type: 'one-for-one' | 'all-for-one';
  maxRetries: number;
  withinMs: number;
  decider: Decider;
}

// ============================================
// 기본 전략 생성 헬퍼
// ============================================

function oneForOneStrategy(
  maxRetries: number,
  withinMs: number,
  decider: Decider
): SupervisorStrategy {
  return { type: 'one-for-one', maxRetries, withinMs, decider };
}

function allForOneStrategy(
  maxRetries: number,
  withinMs: number,
  decider: Decider
): SupervisorStrategy {
  return { type: 'all-for-one', maxRetries, withinMs, decider };
}

// ============================================
// 재시작 통계
// ============================================

class RestartStatistics {
  private failures: number[] = [];

  constructor(
    private maxRetries: number,
    private withinMs: number
  ) {}

  recordFailure(): boolean {
    const now = Date.now();
    this.failures.push(now);

    // 시간 범위 내의 실패만 유지
    this.failures = this.failures.filter(t => now - t < this.withinMs);

    return this.failures.length <= this.maxRetries;
  }

  reset(): void {
    this.failures = [];
  }
}

// ============================================
// Actor 기본 클래스
// ============================================

type ActorRef = SupervisedActor;

abstract class SupervisedActor {
  protected name: string;
  protected children: Map<string, SupervisedActor> = new Map();
  protected parent?: SupervisedActor;
  protected strategy?: SupervisorStrategy;
  protected restartStats: Map<string, RestartStatistics> = new Map();
  protected isRunning = true;

  constructor(name: string, strategy?: SupervisorStrategy) {
    this.name = name;
    this.strategy = strategy;
  }

  // 자식 Actor 생성
  spawn(name: string, actor: SupervisedActor): ActorRef {
    actor.parent = this;
    this.children.set(name, actor);

    if (this.strategy) {
      this.restartStats.set(
        name,
        new RestartStatistics(this.strategy.maxRetries, this.strategy.withinMs)
      );
    }

    console.log(`[${this.name}] Spawned child: ${name}`);
    return actor;
  }

  // 메시지 전송
  send(message: any): void {
    if (!this.isRunning) {
      console.log(`[${this.name}] Actor is stopped, message dropped`);
      return;
    }

    try {
      this.receive(message);
    } catch (error) {
      this.handleFailure(error as Error, message);
    }
  }

  protected abstract receive(message: any): void;

  // 생명주기 훅
  protected preStart(): void {
    console.log(`[${this.name}] Starting...`);
  }

  protected postStop(): void {
    console.log(`[${this.name}] Stopped`);
  }

  protected preRestart(reason: Error): void {
    console.log(`[${this.name}] Restarting due to: ${reason.message}`);
    // 기본: 모든 자식 중지
    this.children.forEach(child => child.stop());
    this.postStop();
  }

  protected postRestart(reason: Error): void {
    this.preStart();
    console.log(`[${this.name}] Restarted`);
  }

  // 장애 처리
  private handleFailure(error: Error, message: any): void {
    console.log(`[${this.name}] Failed with: ${error.message}`);

    if (this.parent) {
      this.parent.supervise(this.name, error, message);
    } else {
      // 루트 Actor는 스스로 처리
      console.log(`[${this.name}] Root actor failure, escalating to system`);
      this.stop();
    }
  }

  // 자식 감독
  private supervise(childName: string, error: Error, message: any): void {
    const child = this.children.get(childName);
    if (!child || !this.strategy) return;

    const directive = this.strategy.decider(error);
    console.log(`[${this.name}] Supervising ${childName}: ${directive}`);

    switch (directive) {
      case Directive.Resume:
        console.log(`[${this.name}] Resuming ${childName}`);
        break;

      case Directive.Restart:
        this.handleRestart(childName, child, error);
        break;

      case Directive.Stop:
        this.handleStop(childName, child);
        break;

      case Directive.Escalate:
        if (this.parent) {
          this.parent.supervise(this.name, error, message);
        }
        break;
    }
  }

  private handleRestart(childName: string, child: SupervisedActor, error: Error): void {
    const stats = this.restartStats.get(childName);

    if (stats && !stats.recordFailure()) {
      console.log(`[${this.name}] Max retries exceeded for ${childName}, stopping`);
      this.handleStop(childName, child);
      return;
    }

    if (this.strategy?.type === 'all-for-one') {
      // 모든 자식 재시작
      console.log(`[${this.name}] All-for-one: restarting all children`);
      this.children.forEach((c, name) => {
        c.preRestart(error);
        c.postRestart(error);
      });
    } else {
      // 해당 자식만 재시작
      child.preRestart(error);
      child.postRestart(error);
    }
  }

  private handleStop(childName: string, child: SupervisedActor): void {
    child.stop();
    this.children.delete(childName);
    this.restartStats.delete(childName);
  }

  stop(): void {
    this.isRunning = false;
    this.children.forEach(child => child.stop());
    this.children.clear();
    this.postStop();
  }
}

// ============================================
// 에러 타입 정의
// ============================================

class TransientError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'TransientError';
  }
}

class DatabaseError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'DatabaseError';
  }
}

class FatalError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'FatalError';
  }
}

// ============================================
// Actor 구현
// ============================================

// 워커 Actor
class WorkerActor extends SupervisedActor {
  private jobCount = 0;

  constructor(name: string) {
    super(name);
    this.preStart();
  }

  protected receive(message: any): void {
    if (message.type === 'PROCESS_JOB') {
      this.jobCount++;
      console.log(`[${this.name}] Processing job #${this.jobCount}: ${message.data}`);

      // 시뮬레이션: 랜덤하게 에러 발생
      if (message.data === 'transient_error') {
        throw new TransientError('Temporary network issue');
      }
      if (message.data === 'fatal_error') {
        throw new FatalError('Critical system failure');
      }

      console.log(`[${this.name}] Job #${this.jobCount} completed`);
    }
  }

  protected preRestart(reason: Error): void {
    console.log(`[${this.name}] Saving state before restart... (jobs processed: ${this.jobCount})`);
    super.preRestart(reason);
  }

  protected postRestart(reason: Error): void {
    this.jobCount = 0; // 상태 초기화
    super.postRestart(reason);
  }
}

// 데이터베이스 Actor (Resume 전략)
class DatabaseActor extends SupervisedActor {
  private connectionPool = 5;

  constructor(name: string) {
    super(name);
    this.preStart();
  }

  protected receive(message: any): void {
    if (message.type === 'QUERY') {
      console.log(`[${this.name}] Executing query: ${message.sql}`);

      if (message.sql === 'bad_query') {
        throw new DatabaseError('Query syntax error');
      }

      console.log(`[${this.name}] Query completed`);
    }
  }

  // Resume 시 상태 유지
  protected preRestart(reason: Error): void {
    // 데이터베이스 Actor는 커넥션 풀 유지
    console.log(`[${this.name}] Keeping connection pool: ${this.connectionPool}`);
  }
}

// 워커 감독자
class WorkerSupervisor extends SupervisedActor {
  constructor() {
    super('WorkerSupervisor', oneForOneStrategy(3, 60000, (error) => {
      if (error instanceof TransientError) {
        return Directive.Restart;
      }
      if (error instanceof FatalError) {
        return Directive.Stop;
      }
      return Directive.Escalate;
    }));
    this.preStart();
  }

  protected receive(message: any): void {
    if (message.type === 'DISPATCH') {
      // 자식들에게 작업 분배
      const workers = Array.from(this.children.values());
      if (workers.length > 0) {
        const worker = workers[Math.floor(Math.random() * workers.length)];
        worker.send({ type: 'PROCESS_JOB', data: message.data });
      }
    }
  }

  initWorkers(count: number): void {
    for (let i = 1; i <= count; i++) {
      this.spawn(`Worker-${i}`, new WorkerActor(`Worker-${i}`));
    }
  }
}

// 루트 감독자
class RootSupervisor extends SupervisedActor {
  constructor() {
    super('RootSupervisor', oneForOneStrategy(5, 60000, (error) => {
      console.log(`[RootSupervisor] Deciding for error: ${error.name}`);
      if (error instanceof DatabaseError) {
        return Directive.Resume;
      }
      return Directive.Restart;
    }));
    this.preStart();
  }

  protected receive(message: any): void {
    console.log(`[RootSupervisor] Received: ${JSON.stringify(message)}`);
  }
}

// ============================================
// 실행 예제
// ============================================

async function main() {
  console.log('=== Supervision Pattern Demo ===\n');

  // 감독 트리 구성
  const root = new RootSupervisor();
  const workerSupervisor = root.spawn('WorkerSupervisor', new WorkerSupervisor()) as WorkerSupervisor;
  const database = root.spawn('Database', new DatabaseActor('Database'));

  // 워커 초기화
  workerSupervisor.initWorkers(3);

  console.log('\n--- 1. Normal Operation ---\n');

  workerSupervisor.send({ type: 'DISPATCH', data: 'job-1' });
  workerSupervisor.send({ type: 'DISPATCH', data: 'job-2' });
  database.send({ type: 'QUERY', sql: 'SELECT * FROM users' });

  await delay(100);

  console.log('\n--- 2. Transient Error (Restart) ---\n');

  workerSupervisor.send({ type: 'DISPATCH', data: 'transient_error' });

  await delay(100);

  console.log('\n--- 3. Database Error (Resume) ---\n');

  database.send({ type: 'QUERY', sql: 'bad_query' });

  await delay(100);

  console.log('\n--- 4. Normal Operation After Recovery ---\n');

  workerSupervisor.send({ type: 'DISPATCH', data: 'job-3' });
  database.send({ type: 'QUERY', sql: 'SELECT * FROM orders' });

  await delay(100);

  console.log('\n--- 5. Fatal Error (Stop) ---\n');

  workerSupervisor.send({ type: 'DISPATCH', data: 'fatal_error' });

  await delay(100);

  console.log('\n--- 6. System Shutdown ---\n');

  root.stop();

  console.log('\n=== Demo Complete ===');
}

function delay(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

main().catch(console.error);

File: examples/03-supervision/csharp/Supervision.cs
// Supervision Pattern - 감독 트리 예제 in C#
// Orleans/Akka.NET 스타일로 구현

using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace SupervisionExample
{
    // ============================================
    // 감독 전략 정의
    // ============================================

    public enum Directive
    {
        Resume,     // 상태 유지하고 계속
        Restart,    // 상태 초기화 후 재시작
        Stop,       // 종료
        Escalate    // 부모에게 전파
    }

    public delegate Directive Decider(Exception error);

    public class SupervisorStrategy
    {
        public string Type { get; init; } = "one-for-one";
        public int MaxRetries { get; init; } = 3;
        public TimeSpan WithinDuration { get; init; } = TimeSpan.FromMinutes(1);
        public Decider Decider { get; init; } = _ => Directive.Escalate;
    }

    // ============================================
    // 재시작 통계
    // ============================================

    public class RestartStatistics
    {
        private readonly List<DateTime> _failures = new();
        private readonly int _maxRetries;
        private readonly TimeSpan _withinDuration;
        private readonly object _lock = new();

        public RestartStatistics(int maxRetries, TimeSpan withinDuration)
        {
            _maxRetries = maxRetries;
            _withinDuration = withinDuration;
        }

        public bool RecordFailure()
        {
            lock (_lock)
            {
                var now = DateTime.UtcNow;
                _failures.Add(now);

                // 시간 범위 내의 실패만 유지
                _failures.RemoveAll(t => now - t > _withinDuration);

                return _failures.Count <= _maxRetries;
            }
        }

        public void Reset()
        {
            lock (_lock)
            {
                _failures.Clear();
            }
        }
    }

    // ============================================
    // 에러 타입 정의
    // ============================================

    public class TransientException : Exception
    {
        public TransientException(string message) : base(message) { }
    }

    public class DatabaseException : Exception
    {
        public DatabaseException(string message) : base(message) { }
    }

    public class FatalException : Exception
    {
        public FatalException(string message) : base(message) { }
    }

    // ============================================
    // Actor 기본 클래스
    // ============================================

    public abstract class SupervisedActor
    {
        protected readonly string Name;
        protected readonly Dictionary<string, SupervisedActor> Children = new();
        protected SupervisedActor? Parent;
        protected readonly SupervisorStrategy? Strategy;
        protected readonly Dictionary<string, RestartStatistics> RestartStats = new();
        protected bool IsRunning = true;
        protected readonly object Lock = new();

        protected SupervisedActor(string name, SupervisorStrategy? strategy = null)
        {
            Name = name;
            Strategy = strategy;
        }

        public SupervisedActor Spawn(string name, SupervisedActor child)
        {
            lock (Lock)
            {
                child.Parent = this;
                Children[name] = child;

                if (Strategy != null)
                {
                    RestartStats[name] = new RestartStatistics(
                        Strategy.MaxRetries,
                        Strategy.WithinDuration);
                }

                Console.WriteLine($"[{Name}] Spawned child: {name}");
                child.PreStart();
                return child;
            }
        }

        public void Send(object message)
        {
            if (!IsRunning)
            {
                Console.WriteLine($"[{Name}] Actor is stopped, message dropped");
                return;
            }

            try
            {
                Receive(message);
            }
            catch (Exception ex)
            {
                HandleFailure(ex, message);
            }
        }

        protected abstract void Receive(object message);

        protected virtual void PreStart()
        {
            Console.WriteLine($"[{Name}] Starting...");
        }

        protected virtual void PostStop()
        {
            Console.WriteLine($"[{Name}] Stopped");
        }

        protected virtual void PreRestart(Exception reason)
        {
            Console.WriteLine($"[{Name}] Restarting due to: {reason.Message}");
            foreach (var child in Children.Values)
            {
                child.Stop();
            }
            PostStop();
        }

        protected virtual void PostRestart(Exception reason)
        {
            PreStart();
            Console.WriteLine($"[{Name}] Restarted");
        }

        private void HandleFailure(Exception error, object message)
        {
            Console.WriteLine($"[{Name}] Failed with: {error.Message}");

            if (Parent != null)
            {
                Parent.Supervise(Name, error, message);
            }
            else
            {
                Console.WriteLine($"[{Name}] Root actor failure");
                Stop();
            }
        }

        protected void Supervise(string childName, Exception error, object message)
        {
            SupervisedActor? child;
            lock (Lock)
            {
                if (!Children.TryGetValue(childName, out child) || Strategy == null)
                    return;
            }

            var directive = Strategy.Decider(error);
            Console.WriteLine($"[{Name}] Supervising {childName}: {directive}");

            switch (directive)
            {
                case Directive.Resume:
                    Console.WriteLine($"[{Name}] Resuming {childName}");
                    break;

                case Directive.Restart:
                    HandleRestart(childName, child, error);
                    break;

                case Directive.Stop:
                    HandleStop(childName, child);
                    break;

                case Directive.Escalate:
                    Parent?.Supervise(Name, error, message);
                    break;
            }
        }

        private void HandleRestart(string childName, SupervisedActor child, Exception error)
        {
            RestartStatistics? stats;
            lock (Lock)
            {
                RestartStats.TryGetValue(childName, out stats);
            }

            if (stats != null && !stats.RecordFailure())
            {
                Console.WriteLine($"[{Name}] Max retries exceeded for {childName}, stopping");
                HandleStop(childName, child);
                return;
            }

            if (Strategy?.Type == "all-for-one")
            {
                Console.WriteLine($"[{Name}] All-for-one: restarting all children");
                lock (Lock)
                {
                    foreach (var c in Children.Values)
                    {
                        c.PreRestart(error);
                        c.PostRestart(error);
                    }
                }
            }
            else
            {
                child.PreRestart(error);
                child.PostRestart(error);
            }
        }

        private void HandleStop(string childName, SupervisedActor child)
        {
            child.Stop();
            lock (Lock)
            {
                Children.Remove(childName);
                RestartStats.Remove(childName);
            }
        }

        public void Stop()
        {
            List<SupervisedActor> childrenToStop;
            lock (Lock)
            {
                IsRunning = false;
                childrenToStop = new List<SupervisedActor>(Children.Values);
                Children.Clear();
            }

            foreach (var child in childrenToStop)
            {
                child.Stop();
            }
            PostStop();
        }
    }

    // ============================================
    // Actor 구현
    // ============================================

    public class WorkerActor : SupervisedActor
    {
        private int _jobCount = 0;

        public WorkerActor(string name) : base(name) { }

        protected override void Receive(object message)
        {
            if (message is ProcessJob job)
            {
                _jobCount++;
                Console.WriteLine($"[{Name}] Processing job #{_jobCount}: {job.Data}");

                if (job.Data == "transient_error")
                {
                    throw new TransientException("Temporary network issue");
                }
                if (job.Data == "fatal_error")
                {
                    throw new FatalException("Critical system failure");
                }

                Console.WriteLine($"[{Name}] Job #{_jobCount} completed");
            }
        }

        protected override void PreRestart(Exception reason)
        {
            Console.WriteLine($"[{Name}] Saving state before restart... (jobs processed: {_jobCount})");
            base.PreRestart(reason);
        }

        protected override void PostRestart(Exception reason)
        {
            _jobCount = 0;
            base.PostRestart(reason);
        }
    }

    public class DatabaseActor : SupervisedActor
    {
        private readonly int _connectionPool = 5;

        public DatabaseActor(string name) : base(name) { }

        protected override void Receive(object message)
        {
            if (message is QueryMessage query)
            {
                Console.WriteLine($"[{Name}] Executing query: {query.Sql}");

                if (query.Sql == "bad_query")
                {
                    throw new DatabaseException("Query syntax error");
                }

                Console.WriteLine($"[{Name}] Query completed");
            }
        }

        protected override void PreRestart(Exception reason)
        {
            Console.WriteLine($"[{Name}] Keeping connection pool: {_connectionPool}");
        }
    }

    public class WorkerSupervisor : SupervisedActor
    {
        private readonly Random _random = new();

        public WorkerSupervisor() : base("WorkerSupervisor", new SupervisorStrategy
        {
            Type = "one-for-one",
            MaxRetries = 3,
            WithinDuration = TimeSpan.FromMinutes(1),
            Decider = error => error switch
            {
                TransientException => Directive.Restart,
                FatalException => Directive.Stop,
                _ => Directive.Escalate
            }
        })
        { }

        protected override void Receive(object message)
        {
            if (message is DispatchJob dispatch)
            {
                List<SupervisedActor> workers;
                lock (Lock)
                {
                    workers = new List<SupervisedActor>(Children.Values);
                }

                if (workers.Count > 0)
                {
                    var worker = workers[_random.Next(workers.Count)];
                    worker.Send(new ProcessJob(dispatch.Data));
                }
            }
        }

        public void InitWorkers(int count)
        {
            for (int i = 1; i <= count; i++)
            {
                var name = $"Worker-{i}";
                Spawn(name, new WorkerActor(name));
            }
        }
    }

    public class RootSupervisor : SupervisedActor
    {
        public RootSupervisor() : base("RootSupervisor", new SupervisorStrategy
        {
            Type = "one-for-one",
            MaxRetries = 5,
            WithinDuration = TimeSpan.FromMinutes(1),
            Decider = error =>
            {
                Console.WriteLine($"[RootSupervisor] Deciding for error type");
                return error switch
                {
                    DatabaseException => Directive.Resume,
                    _ => Directive.Restart
                };
            }
        })
        { }

        protected override void Receive(object message)
        {
            Console.WriteLine($"[{Name}] Received: {message}");
        }
    }

    // ============================================
    // 메시지 타입
    // ============================================

    public record ProcessJob(string Data);
    public record DispatchJob(string Data);
    public record QueryMessage(string Sql);

    // ============================================
    // 메인 프로그램
    // ============================================

    public class Program
    {
        public static async Task Main()
        {
            Console.WriteLine("=== Supervision Pattern Demo (C#) ===\n");

            // 감독 트리 구성
            var root = new RootSupervisor();
            root.Send(null!); // PreStart trigger

            var workerSupervisor = (WorkerSupervisor)root.Spawn("WorkerSupervisor", new WorkerSupervisor());
            var database = root.Spawn("Database", new DatabaseActor("Database"));

            workerSupervisor.InitWorkers(3);

            Console.WriteLine("\n--- 1. Normal Operation ---\n");

            workerSupervisor.Send(new DispatchJob("job-1"));
            workerSupervisor.Send(new DispatchJob("job-2"));
            database.Send(new QueryMessage("SELECT * FROM users"));

            await Task.Delay(100);

            Console.WriteLine("\n--- 2. Transient Error (Restart) ---\n");

            workerSupervisor.Send(new DispatchJob("transient_error"));

            await Task.Delay(100);

            Console.WriteLine("\n--- 3. Database Error (Resume) ---\n");

            database.Send(new QueryMessage("bad_query"));

            await Task.Delay(100);

            Console.WriteLine("\n--- 4. Normal Operation After Recovery ---\n");

            workerSupervisor.Send(new DispatchJob("job-3"));
            database.Send(new QueryMessage("SELECT * FROM orders"));

            await Task.Delay(100);

            Console.WriteLine("\n--- 5. Fatal Error (Stop) ---\n");

            workerSupervisor.Send(new DispatchJob("fatal_error"));

            await Task.Delay(100);

            Console.WriteLine("\n--- 6. System Shutdown ---\n");

            root.Stop();

            Console.WriteLine("\n=== Demo Complete ===");
        }
    }
}
